src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -303,33 +298,28 @@ private DecisionTree AddByType(DecisionTree.ByType byType, TypeSymbol type, Deci
                     }
                 }
             }
-            foreach (var kvp in byType.TypeAndDecision)
+
+            // if the last type is the type we need, add to it
+            DecisionTree result = null;
+            if (byType.TypeAndDecision.Count != 0)
             {
-                var MatchedType = kvp.Key;
-                var Decision = kvp.Value;
-                // See if matching Type matches this value
-                switch (ExpressionOfTypeMatchesPatternType(type, MatchedType, ref _useSiteDiagnostics))
+                var lastTypeAndDecision = byType.TypeAndDecision.Last();
+                if (lastTypeAndDecision.Key.TupleUnderlyingTypeOrSelf() == type.TupleUnderlyingTypeOrSelf())
```

Comment:
I think you misunderstand the intent. The code isn't intended to adapt two types that differ in those ways. It is intended to treat tuple types as their underlying types, because the binder turns a constructed type into a tuple type in some circumstances, and then test for type equality. Whether relaxing that to account for differences in custom modifiers (which as far as I know cannot be designated in source, and therefore cannot appear in this map) and nested tuple names would produce correct (and perhaps slightly better) code in some circumstances is an open question that I am not investigating at this time.

---
In reply to: [100616402](https://github.com/dotnet/roslyn/pull/16928#discussion_r100616402) [](ancestors = 100616402,100602791,100601157,100591895)